pgb_utils/pgb_utils/bigquery.py

```diff
@@ -0,0 +1,634 @@
+    """
+    dflist = []
+    for r, row in enumerate(query_job):
+        # convert to DataFrame
+        df = _format_history_row_to_df(row)
+        # add the objectId so we can use it to multi-index
+        df['objectId'] = df.objectId
+        # set the multi-index and append to the list
+        dflist.append(df.reset_index().set_index(['objectId','candid']))
+
+    histories = pd.concat(dflist)
+
+    return histories
+
+def _format_history_row_to_df(row: Union[dict,bigquery.table.Row]):
+    """Convert a single object's history from a query row to a DataFrame.
+    Any duplicate observations will be dropped.
+    """
+    d = dict(row.items())
+    oid, cid = d.pop('objectId'), d.pop('candid')
+    df = pd.DataFrame(data=d, index=pd.Index(cid, name='candid'))
+    df.drop_duplicates(inplace=True)
+    df.objectId = oid
+    return df
+
+
+#--- Cone Search
+def cone_search(center: astropy.coordinates.SkyCoord,
+                radius: astropy.coordinates.Angle,
+                columns: List[str],
+                objectIds: Optional[list] = None,
+                format: str = 'pandas',
+                iterator: bool = False,
+                dry_run: bool = True,
+                ) -> Union[str,
+                           pd.DataFrame,
+                           Generator[Union[str, pd.DataFrame], None, None]]:
+    """Perform a cone search on the alerts database and return object histories.
+    This uses the coordinates of the most recent observation to determine
+    whether an object is within the cone.
+
+    Args:
+        center: Center of the cone to search within.
+        radius: Radius of the cone to search within.
+        columns: Names of history columns to select from the alerts table.
+                 The 'objectId' and 'candid' columns are automatically included
+                 and do not need to be in this list.
+        objectIds: IDs of ZTF objects to include in the query.
+        format: One of 'pandas', or 'json'. Query results will be
+                returned in this format. Duplicate observations are dropped.
+        iterator: If True, iterate over the objects and return one at a time.
+                  Else return the full query results together.
+        dry_run: If True, `pgb.bigquery.dry_run` will be called first and the
+                 user will be asked to confirm before continuing.
+
+    Returns: Query results in the requested format. If `iterator` is True,
+             returns a generator; else all results are returned together.
+    """
+    # make sure we have required columns
+    for c in ['jd', 'ra', 'dec']:
+        if c not in columns: columns.append(c)
+
+    # Performing a dry run prints the SQL query statement, which does not account
+    # for the cone search. We'll print some things to reduce user confusion.
+    if dry_run: print('\nInitiating a cone search.')
+
+    # Query the database for object histories.
+    objects = query_objects(columns,
```

Comment:
Can you point me to a reference to find the pseudo-rectangular limits? I would guess to look in the ZTF instrument paper? I think I can guess what this mean, but that's a new term to me.

Reply:
I just mean some bounds: (min RA, max RA), (min Dec, max Dec).

I just made up the term "pseudo-rectangular" because it's not a rectangle.

The one non-trivial part is that you have to handle ra wrap-around correctly.

Reply:
I opened issue #50 for this.